server/src/main/java/io/envoyproxy/controlplane/server/DiscoveryServerCallbacks.java

```diff
@@ -0,0 +1,68 @@
+package io.envoyproxy.controlplane.server;
+
+import envoy.api.v2.Discovery.DiscoveryRequest;
+import envoy.api.v2.Discovery.DiscoveryResponse;
+
+/**
+ * {@code DiscoveryServerCallbacks} defines the callbacks that are exposed by the {@link DiscoveryServer}. The callbacks
+ * give consumers the opportunity to add their own application-specific logs/metrics based
+ */
+public interface DiscoveryServerCallbacks {
+
+  /**
+   * {@code onStreamClose} is called just before the bi-directional gRPC stream is closed successfully for an envoy
+   * instance.
+   *
+   * @param streamId an ID for this stream that is only unique to this discovery server instance
+   * @param typeUrl the resource type of the stream, or {@link DiscoveryServer#ANY_TYPE_URL} for ADS
+   */
+  default void onStreamClose(long streamId, String typeUrl) {
```

Comment:
thinking out loud: maybe we should be passing around an enum instead of the typeUrl? avoid having to do string comparisons all over the place. might be out of the scope of this PR though, i think we do this elsewhere already

Reply:
I originally did use enums, but it was getting really messy because we were constantly converting back and forth between the enum type and the underlying string representation.

Once the dust settles after all this major refactoring, I'm definitely open to revisiting that.

Reply:
I think it would be beneficial to pass the `Node` information to these callbacks along with tyleUl

Reply:
I agree that would be useful information, but with the current approach we don't really know the `Node` value when the stream is opened/closed due to how the stream observers are designed. In the `onStreamOpen` scenario, the first request hasn't actually been sent yet, and for `onStreamClose` there's no associated message to pull the node from.

Some ideas for how to work around that:
- Delay the `onStreamOpen` call until the first `onNext(request)` is handled. We would need to maintain some additional state to track whether or not that callback had been executed yet for the current observer instance. I don't think there are enough interesting things that happen between the stream opening and the first request in this scenario for us to care that the timing for `onStreamOpen` is a little delayed.
- Store a copy of the last `request.getNode()` value in the observer instance so that `onError` and `onCompleted` can reference it. We would need the extra state for tracking the last node and consumers of `onStreamClose` and `onStreamCloseWithError` would need to be aware that the `Node` param is `@Nullable` because the stream may close before we ever received a message.

I'm not opposed to making those changes, but I just wanted to open it up for discussion since it's not just a trivial addition.

Reply:
Let us track that as a new issue and merge this one.